File: android/app/src/main/java/com/fitnessrider/audio/TapTempoDetector.kt
package com.fitnessrider.audio

import kotlin.math.roundToInt

class TapTempoDetector(
    private val maxHistory: Int = 8,
    private val resetThresholdMs: Long = 2500L
) {
    private val tapTimestamps = mutableListOf<Long>()

    val tapCount: Int
        get() = tapTimestamps.size

    /**
     * Record a tap at the given timestamp (defaults to current system time).
     * Returns the calculated BPM if at least 2 taps are recorded, or null if insufficient taps.
     */
    fun recordTap(now: Long = System.currentTimeMillis()): Double? {
        if (tapTimestamps.isNotEmpty()) {
            val elapsedSinceLast = now - tapTimestamps.last()
            if (elapsedSinceLast > resetThresholdMs) {
                tapTimestamps.clear()
            }
        }

        tapTimestamps.add(now)
        if (tapTimestamps.size > maxHistory) {
            tapTimestamps.removeAt(0)
        }

        return calculateCurrentBpm()
    }

    /**
     * Calculates the rolling average BPM from the recorded timestamps.
     */
    fun calculateCurrentBpm(): Double? {
        if (tapTimestamps.size < 2) return null

        val intervals = mutableListOf<Long>()
        for (i in 1 until tapTimestamps.size) {
            val delta = tapTimestamps[i] - tapTimestamps[i - 1]
            if (delta in 200..2000) { // Valid interval between 30 BPM and 300 BPM
                intervals.add(delta)
            }
        }

        if (intervals.isEmpty()) return null

        val avgIntervalMs = intervals.average()
        if (avgIntervalMs <= 0.0) return null

        val rawBpm = 60_000.0 / avgIntervalMs
        // Round to 1 decimal place
        return ((rawBpm * 10.0).roundToInt() / 10.0).coerceIn(40.0, 240.0)
    }

    fun reset() {
        tapTimestamps.clear()
    }

    companion object {
        fun calculateBpmFromIntervals(intervalsMs: List<Long>): Double? {
            if (intervalsMs.isEmpty()) return null
            val valid = intervalsMs.filter { it in 200..2000 }
            if (valid.isEmpty()) return null
            val avg = valid.average()
            val rawBpm = 60_000.0 / avg
            return ((rawBpm * 10.0).roundToInt() / 10.0).coerceIn(40.0, 240.0)
        }
    }
}

File: ios/FitnessRider/Audio/TapTempoDetector.swift
import Foundation

public final class TapTempoDetector: @unchecked Sendable {
    private var tapTimestamps: [TimeInterval] = []
    private let maxHistory: Int
    private let resetThreshold: TimeInterval

    public init(maxHistory: Int = 8, resetThreshold: TimeInterval = 2.5) {
        self.maxHistory = maxHistory
        self.resetThreshold = resetThreshold
    }

    public var tapCount: Int {
        return tapTimestamps.count
    }

    @discardableResult
    public func recordTap(at time: TimeInterval = Date().timeIntervalSince1970) -> Double? {
        if let last = tapTimestamps.last {
            let elapsed = time - last
            if elapsed > resetThreshold {
                tapTimestamps.removeAll()
            }
        }

        tapTimestamps.append(time)
        if tapTimestamps.count > maxHistory {
            tapTimestamps.removeFirst()
        }

        return calculateCurrentBpm()
    }

    public func calculateCurrentBpm() -> Double? {
        guard tapTimestamps.count >= 2 else { return nil }

        var intervals: [TimeInterval] = []
        for i in 1..<tapTimestamps.count {
            let delta = tapTimestamps[i] - tapTimestamps[i - 1]
            if delta >= 0.2 && delta <= 2.0 {
                intervals.append(delta)
            }
        }

        guard !intervals.isEmpty else { return nil }
        let avgInterval = intervals.reduce(0.0, +) / Double(intervals.count)
        guard avgInterval > 0.0 else { return nil }

        let rawBpm = 60.0 / avgInterval
        let clamped = min(240.0, max(40.0, rawBpm))
        return (clamped * 10.0).rounded() / 10.0
    }

    public func reset() {
        tapTimestamps.removeAll()
    }

    public static func calculateBpm(from intervals: [TimeInterval]) -> Double? {
        let valid = intervals.filter { $0 >= 0.2 && $0 <= 2.0 }
        guard !valid.isEmpty else { return nil }
        let avg = valid.reduce(0.0, +) / Double(valid.count)
        guard avg > 0.0 else { return nil }
        let rawBpm = 60.0 / avg
        let clamped = min(240.0, max(40.0, rawBpm))
        return (clamped * 10.0).rounded() / 10.0
    }
}
